packages/url/src/index.js

```diff
@@ -50,7 +58,7 @@ export function isEmail( email ) {
  * const protocol2 = getProtocol( 'https://wordpress.org' ); // 'https:'
  * ```
  *
- * @return {?string} The protocol part of the URL.
+ * @return {string|void} The protocol part of the URL.
```

Comment:
Regarding the return types here in general, it's been bothering me a bit that some are `SomeType | void` and another is `SomeType | undefined`…

It appears that `type void = undefined | null` or with `strictNullChecks` it's `type void = undefined` (our case). However, when I think of functions that return `void`, I tend to think of side-effects. Using `void` in a union with another type to indicate we may or may not get a value seems like something better expressed by `SomeType | undefined`.

As far as I can tell and for our purposes, `void` and `undefined` are type aliases, so this is very much a stylistic preference, but is this something you considered and do you have thoughts on it?

Reply:
I noticed that in the TypeScript playground, a `return` statement —even empty `return;`— is required to return `undefined` instead of `void`. Viewed from this perspective and accustomed to JavaScript's implicit `undefined` return, that seems like a strange decision 🤷‍♂ 

It's tempting to go add empty returns to the functions that are returning `string | void`, but TypeScript should be supporting us and not the other way around.

These are fine to leave, thanks for your thoughts 👍

Reply:
My only worry is that we will have to update all occurrences where the return value is optional. It might be quite a huge number of lines. For what it's worth, it's at list very explicit.